File: crates/paperjam-cli/src/commands/sanitize.rs
use paperjam_core::sanitize::SanitizeOptions;

use crate::cli::{Cli, OutputFormat, SanitizeArgs};
use crate::document::open_document;
use crate::error::CliError;

pub fn run(args: &SanitizeArgs, cli: &Cli) -> Result<(), CliError> {
    let doc = open_document(&args.file, cli.password.as_deref())?;

    let options = SanitizeOptions {
        remove_javascript: !args.keep_javascript,
        remove_embedded_files: !args.keep_embedded_files,
        remove_actions: !args.keep_actions,
        remove_links: !args.keep_links,
    };

    let (mut new_doc, result) = paperjam_core::sanitize::sanitize(&doc, &options)?;
    new_doc.save(&args.output)?;

    match cli.format {
        OutputFormat::Json => {
            let json = serde_json::json!({
                "javascript_removed": result.javascript_removed,
                "embedded_files_removed": result.embedded_files_removed,
                "actions_removed": result.actions_removed,
                "links_removed": result.links_removed,
            });
            println!("{}", serde_json::to_string_pretty(&json).unwrap());
        }
        OutputFormat::Text => {
            if !cli.quiet {
                let total = result.javascript_removed
                    + result.embedded_files_removed
                    + result.actions_removed
                    + result.links_removed;
                println!(
                    "Sanitized: {} item(s) removed -> {}",
                    total,
                    args.output.display()
                );
                if result.javascript_removed > 0 {
                    println!("  JavaScript: {}", result.javascript_removed);
                }
                if result.embedded_files_removed > 0 {
                    println!("  Embedded files: {}", result.embedded_files_removed);
                }
                if result.actions_removed > 0 {
                    println!("  Actions: {}", result.actions_removed);
                }
                if result.links_removed > 0 {
                    println!("  Links: {}", result.links_removed);
                }
            }
        }
    }

    Ok(())
}

File: examples/security/sanitize.py
"""Sanitize a PDF by removing JavaScript, embedded files, and dangerous actions."""

import argparse
from pathlib import Path

import paperjam


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Sanitize a PDF by removing potentially dangerous content.",
    )
    parser.add_argument("input", help="Path to the input PDF")
    parser.add_argument(
        "-o",
        "--output",
        default="./output",
        help="Output directory (default: ./output)",
    )
    parser.add_argument(
        "--keep-js",
        action="store_true",
        help="Keep JavaScript (don't remove)",
    )
    parser.add_argument(
        "--keep-files",
        action="store_true",
        help="Keep embedded files",
    )
    parser.add_argument(
        "--keep-actions",
        action="store_true",
        help="Keep auto-launch actions",
    )
    parser.add_argument(
        "--keep-links",
        action="store_true",
        help="Keep link annotations",
    )
    args = parser.parse_args()

    output = Path(args.output)
    output.mkdir(parents=True, exist_ok=True)

    doc = paperjam.open_pdf(args.input)
    print(f"Opened: {args.input} ({doc.page_count} pages)")

    sanitized, result = doc.sanitize(
        remove_javascript=not args.keep_js,
        remove_embedded_files=not args.keep_files,
        remove_actions=not args.keep_actions,
        remove_links=not args.keep_links,
    )

    print("\nSanitization results:")
    print(f"  JavaScript removed:      {result.javascript_removed}")
    print(f"  Embedded files removed:  {result.embedded_files_removed}")
    print(f"  Actions removed:         {result.actions_removed}")
    print(f"  Links removed:           {result.links_removed}")
    print(f"  Total items removed:     {result.total_removed}")

    if result.items:
        print("\nDetails:")
        for item in result.items:
            loc = f" (page {item.page})" if item.page else ""
            print(f"  [{item.category}]{loc} {item.description}")

    out = output / f"sanitized_{Path(args.input).name}"
    sanitized.save(str(out))
    print(f"\nSaved sanitized PDF to {out}")


if __name__ == "__main__":
    main()
